### ISSUE TITLE
[Docs] How to draw a line in the Poincaré disk with geoopt?

### ISSUE THREAD

**Issue by lancejpollard** [geoopt/geoopt#207]
If my question on drawing a hyperbolic tessellation https://github.com/geoopt/geoopt/issues/205 is too involved, a simpler thing to do, which ChatGPT still is unable to figure out, is how to draw a simple line on the hyperbolic plane, using the Poincaré disk (I'm guessing the `PoincareBall` class). Any drawing/rendering solution would work, just so I could see how to apply this library for computing a line on the hyperbolic plane (and having the drawing would make it "click" better in my brain).

All I am looking for is a little snippet of code like this below. Here are two versions ChatGPT gave me, but it doesn't appear to know the API perfectly. But this is all I'm really looking for:

```py
import geoopt
import matplotlib.pyplot as plt

# Create PoincareBall manifold
manifold = geoopt.PoincareBall()

# Define a start and end point for the line
start = [1, 0]
end = [-1, 0]

# Use the expmap function to map the start and end points to the manifold
start_exp = manifold.expmap(start)
end_exp = manifold.expmap(end)

# Use the logmap function to find the geodesic path between the start and end points
path = manifold.logmap(start_exp, end_exp)

# Plot the line using matplotlib
plt.plot(path[:, 0], path[:, 1])
plt.show()
```

Or:

```py
import geoopt
import torch
import matplotlib.pyplot as plt

# Create the Poincare ball model
poincare = geoopt.PoincareBall()

# Define two points in the hyperbolic space
point1 = torch.tensor([0.1, 0.2])
point2 = torch.tensor([0.3, 0.4])

#Map the points to the tangent space at the identity element
point1_tangent = poincare.expmap(point1, torch.tensor([1.0,0.0,0.0,1.0]))
point2_tangent = poincare.expmap(point2, torch.tensor([1.0,0.0,0.0,1.0]))

# Map the points back to the hyperbolic space
point1_hyperbolic = poincare.logmap(point1_tangent, torch.tensor([1.0,0.0,0.0,1.0]))
point2_hyperbolic = poincare.logmap(point2_tangent, torch.tensor([1.0,0.0,0.0,1.0]))

# Transform the points using the Poincare ball model
transformed_point1 = poincare.mobius_add(torch.tensor([1.0,0.0,0.0,1.0]), point1_hyperbolic)
transformed_point2 = poincare.mobius_add(torch.tensor([1.0,0.0,0.0,1.0]), point2_hyperbolic)

# Plot the line connecting the two points
plt.plot([transformed_point1[0], transformed_point2[0]], [transformed_point1[1], transformed_point2[1]])
plt.show()
```

Neither worked for various reasons. If there is an even simpler rendering option than `matplotlib` by all means that would be great too! Just looking for how to compute a line in geoopt, then to make sure I computed it right, draw a quick image.

Once I can draw a line, I can perhaps port [that code](https://github.com/geoopt/geoopt/issues/205#issuecomment-1403384552) in the linked issue (from TypeScript to Python), and get it to draw a tessellation! Or so it seems.

Basically to draw a tessellation you need to first draw a polygon (could also show how to draw that instead of a line if you wanted!). Then the polygon gets _reflected_ along each edge, recursively. After I know how to draw a line (a polygon "segment"), then I can compute the polygon, then compute the reflections to get the remaining polygons in the tessellation. Then just plot that! Super excited to get a demo of this going, to see if it's possible. If it is possible using geoopt, that would be really cool and solidify my belief that geoopt is the "way forward" in terms of doing hyperbolic geometry (using the "universal model").

Thank you very much for your time.

**Comment by lancejpollard** [comment]
For example, how did you draw this? What is the code for that?

<img width="458" alt="Screenshot 2023-01-27 at 7 11 25 AM" src="https://user-images.githubusercontent.com/28718/215119954-60cf2354-97fb-4edb-bfb0-b4f1e50ea915.png">

**Comment by ferrine** [comment]
Looks like I did not maintain this piece of the plotting code

**Comment by ferrine** [comment]
What I did was basically creating a grid in the tangent space and applying exp0 to it

**Comment by lancejpollard** [comment]
> I’ll send you the snippet, I remember how I did that!

@ferrine Awesome, thank you! If you want you can paste it here otherwise my email is lancejpollard@gmail.com

Looking forward to it!

**Comment by ferrine** [comment]
```
def add_geodesic_grid(ax: plt.Axes, manifold: geoopt.Stereographic, line_width=0.1):

    # define geodesic grid parameters
    N_EVALS_PER_GEODESIC = 10000
    STYLE = "--"
    COLOR = "gray"
    LINE_WIDTH = line_width

    # get manifold properties
    K = manifold.k.item()
    R = manifold.radius.item()

    # get maximal numerical distance to origin on manifold
    if K < 0:
        # create point on R
        r = torch.tensor((R, 0.0), dtype=manifold.dtype)
        # project point on R into valid range (epsilon border)
        r = manifold.projx(r)
        # determine distance from origin
        max_dist_0 = manifold.dist0(r).item()
    else:
        max_dist_0 = np.pi * R
    # adjust line interval for spherical geometry
    circumference = 2*np.pi*R

    # determine reasonable number of geodesics
    # choose the grid interval size always as if we'd be in spherical
    # geometry, such that the grid interpolates smoothly and evenly
    # divides the sphere circumference
    n_geodesics_per_circumference = 4 * 6  # multiple of 4!
    n_geodesics_per_quadrant = n_geodesics_per_circumference // 2
    grid_interval_size = circumference / n_geodesics_per_circumference
    if K < 0:
        n_geodesics_per_quadrant = int(max_dist_0 / grid_interval_size)

    # create time evaluation array for geodesics
    if K < 0:
        min_t = -1.2*max_dist_0
    else:
        min_t = -circumference/2.0
    t = torch.linspace(min_t, -min_t, N_EVALS_PER_GEODESIC)[:, None]

    # define a function to plot the geodesics
    def plot_geodesic(gv):
        ax.plot(*gv.t().numpy(), STYLE, color=COLOR, linewidth=LINE_WIDTH)

    # define geodesic directions
    u_x = torch.tensor((0.0, 1.0))
    u_y = torch.tensor((1.0, 0.0))

    # add origin x/y-crosshair
    o = torch.tensor((0.0, 0.0))
    if K < 0:
        x_geodesic = manifold.geodesic_unit(t, o, u_x)
        y_geodesic = manifold.geodesic_unit(t, o, u_y)
        plot_geodesic(x_geodesic)
        plot_geodesic(y_geodesic)
    else:
        # add the crosshair manually for the sproj of sphere
        # because the lines tend to get thicker if plotted
        # as done for K<0
        ax.axvline(0, linestyle=STYLE, color=COLOR, linewidth=LINE_WIDTH)
        ax.axhline(0, linestyle=STYLE, color=COLOR, linewidth=LINE_WIDTH)

    # add geodesics per quadrant
    for i in range(1, n_geodesics_per_quadrant):
        i = torch.as_tensor(float(i))
        # determine start of geodesic on x/y-crosshair
        x = manifold.geodesic_unit(i*grid_interval_size, o, u_y)
        y = manifold.geodesic_unit(i*grid_interval_size, o, u_x)

        # compute point on geodesics
        x_geodesic = manifold.geodesic_unit(t, x, u_x)
        y_geodesic = manifold.geodesic_unit(t, y, u_y)

        # plot geodesics
        plot_geodesic(x_geodesic)
        plot_geodesic(y_geodesic)
        if K < 0:
            plot_geodesic(-x_geodesic)
            plot_geodesic(-y_geodesic)
```
